File: Saltarelle.Test/Saltarelle.Test/Demo04.cs
using System;
using System.Collections.Generic;
using jQueryApi;
using System.Html;
using System.Serialization;

namespace Saltarelle.Test
{
	public class Demo04
	{

		List<Currency> currencies;
		List<Movie> movies;
		OrderDetails orderDetails;

		public static void Main()
		{
			jQuery.OnDocumentReady(new Demo04().Attach);
		}
		
		void Attach() {
			jQuery.OnDocumentReady(() => {
				PopulateCurrencies();
				PopulateMovies();
				CreateOrderDetails();

				var contextHelpers = new JsDictionary<string,object>();
				contextHelpers["app"]= new {};
				contextHelpers["movies"]= movies;
				contextHelpers["convertedPrice"]= ConvertedPrice;
				contextHelpers["currencies"]= currencies;
				contextHelpers["selectedTitle"]=SelectedTitle;

				Observable.Templates( "moviePurchaseTemplate", "#moviePurchaseTemplate" );
				Observable.Link("moviePurchaseTemplate", "#moviePurchase", orderDetails, contextHelpers );

				jQuery.Select( "#submitOrder" ).Click(evt=>{
					Window.Alert("You ordered a movie ticket as follows:\n"+
					             Json.Stringify(orderDetails));
				});


			});
		}

		Func<int,int,string> ConvertedPrice{
			get
			{
				Func<int,int,string> cp = (selectedMovie, selectedCurrency)=>{
					var currency = currencies[selectedCurrency];
					return (selectedMovie != -1)?
						currency.Symbol +(movies[selectedMovie].TicketPrice*currency.Rate).ToString():
						"";
				};
				return cp;
			}
		}

		Func<int, string> SelectedTitle{
			get
			{
				Func<int, string> st = (selectedMovie) =>{
					return (selectedMovie!=-1) ? movies[selectedMovie].Title : "";
				};
				return st;
			}
		}

		void PopulateCurrencies(){
			currencies = new List<Currency>(){
				new Currency{ Name="US", Label="US Dollar", Rate= 1.0m, Symbol= "$" },
				new Currency{ Name="EUR", Label="Euro", Rate= 0.95m, Symbol= "€" },
				new Currency{ Name="GB", Label="Pound", Rate= 0.63m, Symbol= "£" }
			};
		}

		void PopulateMovies(){
			movies = new List<Movie>(){
				new Movie {Title= "The Red Violin",TicketPrice=18 },
				new Movie {Title= "The Inheritance",TicketPrice=16.5m },
				new Movie {Title= "The Incredible Hulk",TicketPrice=21.99m },
			};
		}

		void CreateOrderDetails(){
			orderDetails= new OrderDetails();
		}


		[Serializable]
		class Currency{
			public string  Name {get;set;}
			public string  Label {get;set;}
			public decimal Rate {get;set;}
			public string  Symbol {get;set;}
		}


		[Serializable]
		class OrderDetails {
			public OrderDetails(){
				SelectedMovie=-1;
				SelectedCurrency=1;
			}
			public string Name{get;set;}
			public int SelectedMovie{get;set;}
			public int SelectedCurrency {get;set;}
			public string Request {get;set;}
		}

		[Serializable]
		class Movie{
			public string Title{get;set;}
			public decimal TicketPrice {get;set;}
		}
	}
}

File: Saltarelle.Test/MyApp.Web/js/Demo04.js
(function() {
	////////////////////////////////////////////////////////////////////////////////
	// Saltarelle.Test.Demo04.Currency
	var $Saltarelle_$Test_Demo04$Currency = function() {
	};
	$Saltarelle_$Test_Demo04$Currency.createInstance = function() {
		return $Saltarelle_$Test_Demo04$Currency.$ctor();
	};
	$Saltarelle_$Test_Demo04$Currency.$ctor = function() {
		var $this = {};
		$this.name = null;
		$this.label = null;
		$this.rate = 0;
		$this.symbol = null;
		return $this;
	};
	////////////////////////////////////////////////////////////////////////////////
	// Saltarelle.Test.Demo04.Movie
	var $Saltarelle_$Test_Demo04$Movie = function() {
	};
	$Saltarelle_$Test_Demo04$Movie.createInstance = function() {
		return $Saltarelle_$Test_Demo04$Movie.$ctor();
	};
	$Saltarelle_$Test_Demo04$Movie.$ctor = function() {
		var $this = {};
		$this.title = null;
		$this.ticketPrice = 0;
		return $this;
	};
	////////////////////////////////////////////////////////////////////////////////
	// Saltarelle.Test.Demo04.OrderDetails
	var $Saltarelle_$Test_Demo04$OrderDetails = function() {
	};
	$Saltarelle_$Test_Demo04$OrderDetails.createInstance = function() {
		return $Saltarelle_$Test_Demo04$OrderDetails.$ctor();
	};
	$Saltarelle_$Test_Demo04$OrderDetails.$ctor = function() {
		var $this = {};
		$this.name = null;
		$this.selectedMovie = 0;
		$this.selectedCurrency = 0;
		$this.request = null;
		$this.selectedMovie = -1;
		$this.selectedCurrency = 1;
		return $this;
	};
	////////////////////////////////////////////////////////////////////////////////
	// Saltarelle.Test.Demo04
	var $Saltarelle_Test_Demo04 = function() {
		this.$currencies = null;
		this.$movies = null;
		this.$orderDetails = null;
	};
	$Saltarelle_Test_Demo04.prototype = {
		$attach: function() {
			$(Function.mkdel(this, function() {
				this.$populateCurrencies();
				this.$populateMovies();
				this.$createOrderDetails();
				var contextHelpers = {};
				contextHelpers['app'] = {};
				contextHelpers['movies'] = this.$movies;
				contextHelpers['convertedPrice'] = this.get_$convertedPrice();
				contextHelpers['currencies'] = this.$currencies;
				contextHelpers['selectedTitle'] = this.get_$selectedTitle();
				$.templates('moviePurchaseTemplate', '#moviePurchaseTemplate');
				$.link.moviePurchaseTemplate('#moviePurchase', this.$orderDetails, contextHelpers);
				$('#submitOrder').click(Function.mkdel(this, function(evt) {
					window.alert('You ordered a movie ticket as follows:\n' + JSON.stringify(this.$orderDetails));
				}));
			}));
		},
		get_$convertedPrice: function() {
			var cp = Function.mkdel(this, function(selectedMovie, selectedCurrency) {
				var currency = this.$currencies[selectedCurrency];
				return ((selectedMovie !== -1) ? (currency.symbol + (this.$movies[selectedMovie].ticketPrice * currency.rate).toString()) : '');
			});
			return cp;
		},
		get_$selectedTitle: function() {
			var st = Function.mkdel(this, function(selectedMovie) {
				return ((selectedMovie !== -1) ? this.$movies[selectedMovie].title : '');
			});
			return st;
		},
		$populateCurrencies: function() {
			var $t1 = [];
			var $t2 = $Saltarelle_$Test_Demo04$Currency.$ctor();
			$t2.name = 'US';
			$t2.label = 'US Dollar';
			$t2.rate = 1;
			$t2.symbol = '$';
			$t1.add($t2);
			var $t3 = $Saltarelle_$Test_Demo04$Currency.$ctor();
			$t3.name = 'EUR';
			$t3.label = 'Euro';
			$t3.rate = 0.95;
			$t3.symbol = '€';
			$t1.add($t3);
			var $t4 = $Saltarelle_$Test_Demo04$Currency.$ctor();
			$t4.name = 'GB';
			$t4.label = 'Pound';
			$t4.rate = 0.63;
			$t4.symbol = '£';
			$t1.add($t4);
			this.$currencies = $t1;
		},
		$populateMovies: function() {
			var $t1 = [];
			var $t2 = $Saltarelle_$Test_Demo04$Movie.$ctor();
			$t2.title = 'The Red Violin';
			$t2.ticketPrice = 18;
			$t1.add($t2);
			var $t3 = $Saltarelle_$Test_Demo04$Movie.$ctor();
			$t3.title = 'The Inheritance';
			$t3.ticketPrice = 16.5;
			$t1.add($t3);
			var $t4 = $Saltarelle_$Test_Demo04$Movie.$ctor();
			$t4.title = 'The Incredible Hulk';
			$t4.ticketPrice = 21.99;
			$t1.add($t4);
			this.$movies = $t1;
		},
		$createOrderDetails: function() {
			this.$orderDetails = $Saltarelle_$Test_Demo04$OrderDetails.$ctor();
		}
	};
	$Saltarelle_Test_Demo04.main = function() {
		var $t1 = new $Saltarelle_Test_Demo04();
		$(Function.mkdel($t1, $t1.$attach));
	};
	Type.registerClass(null, 'Saltarelle.$Test.Demo04$Currency', $Saltarelle_$Test_Demo04$Currency, Object);
	Type.registerClass(null, 'Saltarelle.$Test.Demo04$Movie', $Saltarelle_$Test_Demo04$Movie, Object);
	Type.registerClass(null, 'Saltarelle.$Test.Demo04$OrderDetails', $Saltarelle_$Test_Demo04$OrderDetails, Object);
	Type.registerClass(global, 'Saltarelle.Test.Demo04', $Saltarelle_Test_Demo04, Object);
	$Saltarelle_Test_Demo04.main();
})();
